nixos/modules/virtualisation/rkt.nix

```diff
@@ -39,26 +47,97 @@ in
   };
 
   config = mkIf cfg.enable {
-    environment.systemPackages = [ pkgs.rkt ];
-
-    systemd.services.rkt = {
-      description = "rkt metadata service";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "network.target" ];
-      serviceConfig = {
-        ExecStart = "${pkgs.rkt}/bin/rkt metadata-service";
+    environment.systemPackages = with pkgs; [ rkt ];
+
+    systemd.services = {
+      rkt-api = {
+        description = "rkt api service";
+        wantedBy = lib.mkIf (!cfg.useSocketActivation) [ "multi-user.target" ];
+        after = [ "network.target" "rkt-api-tcp.socket" ];
+        requires = [ "rkt-api-tcp.socket" ];
+        serviceConfig = {
+          ExecStart = "${pkgs.rkt}/bin/rkt api-service";
+          DynamicUser = true;
+          SupplementaryGroups = [ "rkt" ];
+          ReadWritePaths = dataDir;
+        };
+      };
+
+      rkt-metadata = {
+        description = "rkt metadata service";
+        wantedBy = lib.mkIf (!cfg.useSocketActivation) [ "multi-user.target" ];
+        after = [ "network.target" "rkt-metadata.socket" ];
+        requires = [ "rkt-metadata.socket" ];
+        serviceConfig = {
+          ExecStart = "${pkgs.rkt}/bin/rkt metadata-service";
+        };
+      };
+
+      rkt-gc = {
+        description = "rkt garbage collection";
+        startAt = optionalString cfg.gc.automatic cfg.gc.dates;
+        serviceConfig = {
+          Type = "oneshot";
+          ExecStart = "${pkgs.rkt}/bin/rkt gc ${cfg.gc.options}";
+        };
       };
     };
 
-    systemd.services.rkt-gc = {
-      description = "rkt garbage collection";
-      startAt = optionalString cfg.gc.automatic cfg.gc.dates;
-      serviceConfig = {
-        Type = "oneshot";
-        ExecStart = "${pkgs.rkt}/bin/rkt gc ${cfg.gc.options}";
+    systemd.sockets = {
+      rkt-api-tcp = {
+        description = "rkt api service socket";
+        wantedBy = lib.mkIf cfg.useSocketActivation [ "sockets.target" ];
+        partOf = [ "rkt-api.service" ];
+        socketConfig = {
+          ListenStream = [ "127.0.0.1:15441" ] ++ lib.optional config.networking.enableIPv6 "[::1]:15441";
+          Service = "rkt-api.service";
+          BindIPv6Only = lib.mkIf config.networking.enableIPv6 "both";
+        };
+      };
+
+      rkt-metadata = {
+        description = "rkt metadata service socket";
+        wantedBy = lib.mkIf cfg.useSocketActivation [ "sockets.target" ];
+        partOf = [ "rkt-metadata.service" ];
+        socketConfig = {
+          ListenStream = "/run/rkt/metadata-svc.sock";
+          SocketMode = "0660";
+          SocketUser = "root";
+          SocketGroup = "root";
+          RemoveOnStop = true;
+        };
       };
     };
 
+    systemd.tmpfiles.rules = [
```

Comment:
Can you explain why all these files are needed?

Reply:
Upstream wants it

Reply:
Which services creates this usually?

Reply:
I should have been clearer - sorry. In order to be able to actually interact with rkt without being root/using sudo, you need to be a member of the ```rkt``` group.

Without this PR:
```
peter@dolores:~ $ rkt list
list: failed to get pod handles: mkdir /var/lib/rkt/pods: permission denied
```

With it:
```
peter@dolores:~ $ rkt list
UUID	APP	IMAGE NAME	STATE	CREATED	STARTED	NETWORKS
```

Reply:
Could we maybe use the upstream file? `environment.etc."tmpfiles.d/rkt.conf".source = "${rkt}/lib/tmpfiles.d/rkt.conf";`?

Reply:
Services should probably be installed the same way. And `systemd.packages` should probably recognize `tmpfiles.d` (https://github.com/NixOS/nixpkgs/issues/40594).

Reply:
As I understand it, files in ```tmpfiles.d``` are not automatically loaded on NixOS. We need to add them to ```systemd.tmpfiles.rules```.

Reply:
From https://github.com/NixOS/nixpkgs/blob/8a93595550fe72371b75c976d0cf8738ddd33d5a/nixos/modules/system/boot/systemd.nix#L773-L781, it seems like they are.

Reply:
Files that are simply part of the package distribution aren't added - so we need to add it to ```/etc/tmpfiles.d``` like it's done here or alternatively ```builtins.readFile```. In any case, there is no need to re-invent the wheel so I will adjust this to use the file distributes with upstream. We could it it similarly for the units.